doc/file.man
.\" $File: file.man,v 1.65 2007/01/25 21:05:46 christos Exp $
.Dd January 8, 2007
.Dt FILE __CSECTION__
.Os
.Sh NAME
.Nm file
.Nd determine file type
.Sh SYNOPSIS
.Nm
.Op Fl bchikLnNprsvz
.Op Fl f Ar namefile
.Op Fl F Ar separator
.Op Fl m Ar magicfiles
.Ar file
.Nm
.Fl C
.Op Fl m Ar magicfile
.Sh DESCRIPTION
This manual page documents version __VERSION__ of the
.Nm
command.
.Pp
.Nm
tests each argument in an attempt to classify it.
There are three sets of tests, performed in this order:
filesystem tests, magic number tests, and language tests.
The
.Em first
test that succeeds causes the file type to be printed.
.Pp
The type printed will usually contain one of the words
.Em text
(the file contains only
printing characters and a few common control
characters and is probably safe to read on an
.Dv ASCII
terminal),
.Em executable
(the file contains the result of compiling a program
in a form understandable to some 
.Dv UNIX
kernel or another),
or
.Em data
meaning anything else (data is usually 
.Sq binary
or non-printable).
Exceptions are well-known file formats (core files, tar archives)
that are known to contain binary data.
When adding local definitions to
.Pa /etc/magic ,
.Em "preserve these keywords" .
People depend on knowing that all the readable files in a directory
have the word 
.Dq text
printed.
Don't do as Berkeley did and change 
.Dq shell commands text
to 
.Dq shell script .
Note that the file
.Pa __MAGIC__
is built mechanically from a large number of small files in
the subdirectory
.Pa Magdir
in the source distribution of this program.
.Pp
The filesystem tests are based on examining the return from a
.Xr stat 2
system call.
The program checks to see if the file is empty,
or if it's some sort of special file.
Any known file types appropriate to the system you are running on
(sockets, symbolic links, or named pipes (FIFOs) on those systems that
implement them)
are intuited if they are defined in
the system header file
.In sys/stat.h .
.Pp
The magic number tests are used to check for files with data in
particular fixed formats.
The canonical example of this is a binary executable (compiled program)
.Dv a.out
file, whose format is defined in 
.In elf.h ,
.In a.out.h
and possibly
.In exec.h
in the standard include directory.
These files have a 
.Sq "magic number"
stored in a particular place
near the beginning of the file that tells the 
.Dv UNIX operating system
that the file is a binary executable, and which of several types thereof.
The concept of a
.Sq "magic number"
has been applied by extension to data files.
Any file with some invariant identifier at a small fixed
offset into the file can usually be described in this way.
The information identifying these files is read from
.Pa /etc/magic
and the compiled
magic file
.Pa __MAGIC__.mgc ,
or 
.Pa __MAGIC__
if the compiled file does not exist. In addition
.Nm
will look in
.Pa $HOME/.magic.mgc ,
or
.Pa $HOME/.magic
for magic entries.
.Pp
If a file does not match any of the entries in the magic file,
it is examined to see if it seems to be a text file.
ASCII, ISO-8859-x, non-ISO 8-bit extended-ASCII character sets
(such as those used on Macintosh and IBM PC systems),
UTF-8-encoded Unicode, UTF-16-encoded Unicode, and EBCDIC
character sets can be distinguished by the different
ranges and sequences of bytes that constitute printable text
in each set.
If a file passes any of these tests, its character set is reported.
ASCII, ISO-8859-x, UTF-8, and extended-ASCII files are identified
as 
.Dq text
because they will be mostly readable on nearly any terminal;
UTF-16 and EBCDIC are only 
.Dq character data
because, while
they contain text, it is text that will require translation
before it can be read.
In addition,
.Nm
will attempt to determine other characteristics of text-type files.
If the lines of a file are terminated by CR, CRLF, or NEL, instead
of the Unix-standard LF, this will be reported.
Files that contain embedded escape sequences or overstriking
will also be identified.
.Pp
Once
.Nm
has determined the character set used in a text-type file,
it will
attempt to determine in what language the file is written.
The language tests look for particular strings (cf
.In names.h
that can appear anywhere in the first few blocks of a file.
For example, the keyword
.Em .br
indicates that the file is most likely a
.Xr troff 1
input file, just as the keyword 
.Em struct
indicates a C program.
These tests are less reliable than the previous
two groups, so they are performed last.
The language test routines also test for some miscellany
(such as 
.Xr tar 1
archives).
.Pp
Any file that cannot be identified as having been written
in any of the character sets listed above is simply said to be
.Dq data .
.Sh OPTIONS
.Bl -tag -width indent
.It Fl b , -brief
Do not prepend filenames to output lines (brief mode).
.It Fl c , -checking-printout
Cause a checking printout of the parsed form of the magic file.
This is usually used in conjunction with the
.Fl m
flag to debug a new magic file before installing it.
.It Fl C , -compile
Write a
.Pa magic.mgc
output file that contains a pre-parsed version of the magic file.
.It Fl e , -exclude Ar testname
Exclude the test named in
.Ar testname
from the list of tests made to determine the file type. Valid test names
are:
.Bl -tag -width
.It apptype
Check for
.Dv EMX
application type (only on EMX).
.It ascii
Check for various types of ascii files.
.It compress
Don't look for, or inside compressed files.
.It elf
Don't print elf details.
.It fortran
Don't look for fortran sequences inside ascii files.
.It soft
Don't consult magic files.
.It tar
Don't examine tar files.
.It token
Don't look for known tokens inside ascii files.
.It troff
Don't look for troff sequences inside ascii files.
.El
.It Fl f , -files-from Ar namefile
Read the names of the files to be examined from 
.Ar namefile
(one per line) 
before the argument list.
Either 
.Ar namefile
or at least one filename argument must be present;
to test the standard input, use 
.Sq -
as a filename argument.
.It Fl F , -separator Ar separator
Use the specified string as the separator between the filename and the
file result returned. Defaults to 
.Sq \&: .
.It Fl h , -no-dereference
option causes symlinks not to be followed
(on systems that support symbolic links). This is the default if the
environment variable
.Dv POSIXLY_CORRECT
is not defined.
.It Fl i , -mime
Causes the file command to output mime type strings rather than the more
traditional human readable ones. Thus it may say
.Dq text/plain; charset=us-ascii
rather
than
.Dq ASCII text .
In order for this option to work, file changes the way
it handles files recognized by the command itself (such as many of the
text file types, directories etc), and makes use of an alternative
.Dq magic
file.
(See
.Dq FILES
section, below).
.It Fl k , -keep-going
Don't stop at the first match, keep going. Subsequent matches will be
prepended by
.Dq "\[rs]012\- ".
(If you want a newline, see 
.Dq "\-r"
option.)
.It Fl L , -dereference
option causes symlinks to be followed, as the like-named option in
.Xr ls 1
(on systems that support symbolic links).
This is the default if the environment variable
.Dv POSIXLY_CORRECT
is defined.
.It Fl m , -magic-file Ar list
Specify an alternate list of files containing magic numbers.
This can be a single file, or a colon-separated list of files.
If a compiled magic file is found alongside, it will be used instead.
With the 
.Fl i
or 
.Fl "mime"
option, the program adds
.Dq .mime
to each file name.
.It Fl n , -no-buffer
Force stdout to be flushed after checking each file.
This is only useful if checking a list of files.
It is intended to be used by programs that want filetype output from a pipe.
.It Fl N , -no-pad
Don't pad filenames so that they align in the output.
.It Fl p , -preserve-date
On systems that support
.Xr utime 2
or
.Xr utimes 2 ,
attempt to preserve the access time of files analyzed, to pretend that
.Nm
never read them.
.It Fl r , -raw
Don't translate unprintable characters to \eooo.
Normally
.Nm
translates unprintable characters to their octal representation.
.It Fl s , -special-files
Normally,
.Nm
only attempts to read and determine the type of argument files which
.Xr stat 2
reports are ordinary files.
This prevents problems, because reading special files may have peculiar
consequences.
Specifying the
.Fl s
option causes
.Nm
to also read argument files which are block or character special files.
This is useful for determining the filesystem types of the data in raw
disk partitions, which are block special files.
This option also causes
.Nm
to disregard the file size as reported by
.Xr stat 2
since on some systems it reports a zero size for raw disk partitions.
.It Fl v , -version
Print the version of the program and exit.
.It Fl z , -uncompress
Try to look inside compressed files.
.It Fl 0 , -print0
Output a null character
.Sq \e0
after the end of the filename. Nice to
.Xr cut 1
the output. This does not affect the separator which is still printed.
.It Fl -help
Print a help message and exit.
.El
.Sh FILES
.Bl -tag -width __MAGIC__.mime.mgc -compact
.It Pa __MAGIC__.mgc
Default compiled list of magic numbers
.It Pa __MAGIC__
Default list of magic numbers
.It Pa __MAGIC__.mime.mgc
Default compiled list of magic numbers, used to output mime types when
the 
.Fl i
option is specified.
.It Pa __MAGIC__.mime
Default list of magic numbers, used to output mime types when the 
.Fl i
option is specified.
.El
.Sh ENVIRONMENT
The environment variable
.Dv MAGIC
can be used to set the default magic number file name.
If that variable is set, then
.Nm
will not attempt to open
.Pa $HOME/.magic .
.Nm
adds
.Dq .mime
and/or
.Dq .mgc
to the value of this variable as appropriate.
However,
.Pa file
has to exist in order for
.Pa file.mime
to be considered.
The environment variable
.Dv POSIXLY_CORRECT
controls (on systems that support symbolic links), if
.Nm
will attempt to follow symlinks or not. If set, then
.Nm
follows symlink, otherwise it does not. This is also controlled
by the
.Fl L
and
.Fl h
options.
.Sh SEE ALSO
.Xr magic __FSECTION__ ,
.Xr strings 1 ,
.Xr od 1 ,
.Xr hexdump 1
.Sh STANDARDS CONFORMANCE
This program is believed to exceed the System V Interface Definition
of FILE(CMD), as near as one can determine from the vague language
contained therein. 
Its behavior is mostly compatible with the System V program of the same name.
This version knows more magic, however, so it will produce
different (albeit more accurate) output in many cases. 
.\" URL: http://www.opengroup.org/onlinepubs/009695399/utilities/file.html
.Pp
The one significant difference 
between this version and System V
is that this version treats any white space
as a delimiter, so that spaces in pattern strings must be escaped.
For example,
.Bd -literal -offset indent 
>10	string	language impress\ 	(imPRESS data)
.Ed
.Pp
in an existing magic file would have to be changed to
.Bd -literal -offset indent 
>10	string	language\e impress	(imPRESS data)
.Ed
.Pp
In addition, in this version, if a pattern string contains a backslash,
it must be escaped.
For example
.Bd -literal -offset indent 
0	string		\ebegindata	Andrew Toolkit document
.Ed
.Pp
in an existing magic file would have to be changed to
.Bd -literal -offset indent 
0	string		\e\ebegindata	Andrew Toolkit document
.Ed
.Pp
SunOS releases 3.2 and later from Sun Microsystems include a
.Nm 
command derived from the System V one, but with some extensions.
My version differs from Sun's only in minor ways.
It includes the extension of the 
.Sq &
operator, used as,
for example,
.Bd -literal -offset indent 
>16	long&0x7fffffff	>0		not stripped
.Ed
.Sh MAGIC DIRECTORY
The magic file entries have been collected from various sources,
mainly USENET, and contributed by various authors.
Christos Zoulas (address below) will collect additional
or corrected magic file entries.
A consolidation of magic file entries 

debian/changelog
file (4.24-2) unstable; urgency=medium

  * Adding plain magic files for compatibility.
  * Using lintian debhelper to install lintian overrides.
  * Removing watch file.
  * Removing useless whitespaces in changelog file.
  * Adding vcs fields in control file.
  * Upgrading package to debhelper 6.

 -- Daniel Baumann <daniel@debian.org>  Sat,  3 May 2008 14:09:00 +0200

file (4.24-1) experimental; urgency=low

  * New upstream release:
    - The following patches went upstream completely:
      102-magic-remove-bennet, 103-magic-remove-bzip, 104-magic-remove-cgm,
      105-magic-remove-com, 106-magic-remove-emf, 108-magic-remove-quicktime,
      201-magic-update-avi, 203-magic-update-cow, 204-magic-update-cracklib,
      206-magic-update-elf-b, 207-magic-update-elf-c, 208-magic-update-hp,
      209-magic-update-jpeg, 210-magic-update-m4v, 211-magic-update-macintosh,
      213-magic-update-pcp, 214-magic-update-perl, 215-magic-update-sh,
      216-magic-update-spectrum, 301-magic-add-adf, 302-magic-add-apple2,
      303-magic-add-arm, 304-magic-add-avg, 305-magic-add-clarion,
      306-magic-add-dds, 307-magic-add-erlang, 308-magic-add-gedcom,
      309-magic-add-git, 309-magic-add-inform, 310-magic-add-llvm,
      311-magic-add-lua, 312-magic-add-luks, 313-magic-add-lzma,
      314-magic-add-mathcad, 315-magic-add-mdi, 316-magic-add-mercurial,
      317-magic-add-mozilla, 318-magic-add-nlm, and 332-magic-add-svg,
      901-file-elf, 902-file-kfreebsd, and 905-file-segfault.
    - The following patches went upstream almost completely:
      998-doc-manpages and 999-conglomeration.
    - Thanks to Reuben Thomas <rrt@sc3d.org> to merge them.

 -- Daniel Baumann <daniel@debian.org>  Tue,  1 Apr 2008 11:30:00 +0200

file (4.23-2) unstable; urgency=low

  * Applied patch from Neil Williams <codehelp@debian.org> to add crossbuild
    support to debian/rules (Closes: #465111).

 -- Daniel Baumann <daniel@debian.org>  Sun, 10 Feb 2008 20:16:00 +0100

file (4.23-1) unstable; urgency=low

  * New upstream release:
    - Removing '101-magic-remove-awk.dpatch', went upstream.
    - Removing '107-magic-remove-msi.dpatch', went upstream.
    - Removing '202-magic-update-blender.dpatch', went upstream.
    - Removing '205-magic-update-elf-a.dpatch', went upstream.
    - Removing '212-magic-update-os2rexx.dpatch', went upstream.
    - Removing '319-magic-add-ocfs.dpatch', went upstream.

 -- Daniel Baumann <daniel@debian.org>  Sat, 12 Jan 2008 17:23:00 +0100

file (4.21-4) unstable; urgency=low

  * Bumping to new policy.
  * Using new homepage field in control.
  * Using ${binary:Version} in all depends.
  * Added patch from Josh Triplett <josh@freedesktop.org> to add new magic for
    Mozilla XUL fastload files (XUL.mfasl and XPC.mfasl; Closes: #452354).
  * Added patch from Seo Sanghyeon <tinuviel@sparcs.kaist.ac.kr> to add new
    magic for Mercurial bundles (Closes: #451067).
  * Correct wrong applied patch for SNES sound files (Closes: #410847).
  * Added patch from Mathias Brodala <info@noctus.net> to add new magic for
    X11 cursor files (Closes: #451246).

 -- Daniel Baumann <daniel@debian.org>  Sun, 23 Dec 2007 18:09:00 +0100

file (4.21-3) unstable; urgency=low

  * Including extra headers in /usr/include/file (Closes: #397307).
  * Added patch from Eduardo Sabbatella <eduardo_sabbatella@yahoo.com.ar> to
    add magic for SGF Smart Game Format (Closes: #428605).
  * Added patch from Josh Triplett <josh@freedesktop.org> to add magic for git
    bundles (Closes: #430070).
  * Added patch from Mathieu Malaterre <mathieu.malaterre@gmail.com> to extend
    JPEG 2000 magic (Closes: #439005).
  * Added patch from Daniele Sempione <scrows@oziosi.org> to add magic for
    Microsoft Document Imaging Format (.mdi) (Closes: #430548).
  * Added patch from Nicolas Collignon <tsointsoin@gmail.com> to add magics
    for OpenSSH and OpenSSL certificates/key files (Closes: #439537).

 -- Daniel Baumann <daniel@debian.org>  Sun, 26 Aug 2007 09:51:00 +0200

file (4.21-2) unstable; urgency=low

  * Applied patch from Michael Casadevall <sonicmctails@gmail.com> to fix
    FTBFS on hurd (Closes: #433716).

 -- Daniel Baumann <daniel@debian.org>  Thu, 19 Jul 2007 08:27:00 +0200

file (4.21-1) unstable; urgency=low

  * New upstream release.
  * Rediffed 211-magic-update-os2rexx.dpatch.
  * Removed 302-magic-add-ape.dpatch, went upstream.
  * Rediffed 304-magic-add-arm.dpatch.
  * Rediffed 317-magic-add-ocfs.dpatch.
  * Rediffed 318-magic-add-par2.dpatch.
  * Rediffed 322-magic-add-scummvm.dpatch.
  * Rediffed 325-magic-add-subversion.dpatch.
  * Rediffed 326-magic-add-supercollider.dpatch.
  * Rediffed 328-magic-add-xen.dpatch.
  * Removed 999-upstream-reg-startend.dpatch, now included in upstream.
  * Updated 905-file-segfault.dpatch.
  * Rediffed 999-conglomeration.dpatch.

 -- Daniel Baumann <daniel@debian.org>  Fri,  1 Jun 2007 20:29:00 +0200

file (4.20-8) unstable; urgency=low

  * Build a python-magic-dbg package.

 -- Daniel Baumann <daniel@debian.org>  Thu, 17 May 2007 20:40:00 +0200

file (4.20-7) unstable; urgency=medium

  * Replacing work around patch from Ubuntu with proper fix from
    Werner Fink <werner@suse.de> of OpenSuse (CVE-2007-2026).

 -- Daniel Baumann <daniel@debian.org>  Thu, 17 May 2007 19:57:00 +0200

file (4.20-6) unstable; urgency=medium

  * Fixed clean: target in rule (Closes: #424269).
  * Added patch from Anthon van der Neut <anthon@mnt.org> to add magic for LUKS
    (Closes: #394246).
  * Added patch from Jelmer Vernooij <jelmer@samba.org> to add magic for
    Power-Tab.
  * Added patch from Sven Anders <anders@anduras.de> to fix a segfault with -c.
  * Added patch from Kees Cook <kees@ubuntu.com> to work around regex DoS
    (CVE-2007-2026) by disabling the respective magic.
  * Added patch from Noel Torres <tecnico@ejerciciosresueltos.com> to add magic
    for SVG (Closes: #417331).

src/file.c
/*
 * Copyright (c) Ian F. Darwin 1986-1995.
 * Software written by Ian F. Darwin and others;
 * maintained 1995-present by Christos Zoulas and others.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice immediately at the beginning of the file, without modification,
 *    this list of conditions, and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *  
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */
/*
 * file - find type of a file or files - main program.
 */

#include "file.h"
#include "magic.h"

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <sys/types.h>
#include <sys/param.h>	/* for MAXPATHLEN */
#include <sys/stat.h>
#ifdef RESTORE_TIME
# if (__COHERENT__ >= 0x420)
#  include <sys/utime.h>
# else
#  ifdef USE_UTIMES
#   include <sys/time.h>
#  else
#   include <utime.h>
#  endif
# endif
#endif
#ifdef HAVE_UNISTD_H
#include <unistd.h>	/* for read() */
#endif
#ifdef HAVE_LOCALE_H
#include <locale.h>
#endif
#ifdef HAVE_WCHAR_H
#include <wchar.h>
#endif

#ifdef HAVE_GETOPT_H
#include <getopt.h>	/* for long options (is this portable?)*/
#else
#undef HAVE_GETOPT_LONG
#endif

#include <netinet/in.h>		/* for byte swapping */

#include "patchlevel.h"

#ifndef	lint
FILE_RCSID("@(#)$File: file.c,v 1.107 2007/01/25 21:05:46 christos Exp $")
#endif	/* lint */


#ifdef S_IFLNK
#define SYMLINKFLAG "Lh"
#else
#define SYMLINKFLAG ""
#endif

# define USAGE  "Usage: %s [-bcik" SYMLINKFLAG "nNrsvz0] [-e test] [-f namefile] [-F separator] [-m magicfiles] file...\n       %s -C -m magicfiles\n"

#ifndef MAXPATHLEN
#define	MAXPATHLEN	512
#endif

private int 		/* Global command-line options 		*/
	bflag = 0,	/* brief output format	 		*/
	nopad = 0,	/* Don't pad output			*/
	nobuffer = 0,   /* Do not buffer stdout 		*/
	nulsep = 0;	/* Append '\0' to the separator		*/

private const char *magicfile = 0;	/* where the magic is	*/
private const char *default_magicfile = MAGIC;
private const char *separator = ":";	/* Default field separator	*/

private char *progname;		/* used throughout 		*/

private struct magic_set *magic;

private void unwrap(char *);
private void usage(void);
#ifdef HAVE_GETOPT_LONG
private void help(void);
#endif
#if 0
private int byteconv4(int, int, int);
private short byteconv2(int, int, int);
#endif

int main(int, char *[]);
private void process(const char *, int);
private void load(const char *, int);


/*
 * main - parse arguments and handle options
 */
int
main(int argc, char *argv[])
{
	int c, i;
	int action = 0, didsomefiles = 0, errflg = 0;
	int flags = 0;
	char *home, *usermagic;
	struct stat sb;
	static const char hmagic[] = "/.magic";
#define OPTSTRING	"bcCdf:F:hikLm:nNprsvz0"
#ifdef HAVE_GETOPT_LONG
	int longindex;
	static const struct option long_options[] =
	{
		{"version", 0, 0, 'v'},
		{"help", 0, 0, 0},
		{"brief", 0, 0, 'b'},
		{"checking-printout", 0, 0, 'c'},
		{"debug", 0, 0, 'd'},
		{"exclude", 0, 0, 'e' },
		{"files-from", 1, 0, 'f'},
		{"separator", 1, 0, 'F'},
		{"mime", 0, 0, 'i'},
		{"keep-going", 0, 0, 'k'},
#ifdef S_IFLNK
		{"dereference", 0, 0, 'L'},
		{"no-dereference", 0, 0, 'h'},
#endif
		{"magic-file", 1, 0, 'm'},
#if defined(HAVE_UTIME) || defined(HAVE_UTIMES)
		{"preserve-date", 0, 0, 'p'},
#endif
		{"uncompress", 0, 0, 'z'},
		{"raw", 0, 0, 'r'},
		{"no-buffer", 0, 0, 'n'},
		{"no-pad", 0, 0, 'N'},
		{"special-files", 0, 0, 's'},
		{"compile", 0, 0, 'C'},
		{"print0", 0, 0, '0'},
		{0, 0, 0, 0},
	};
#endif

	static const struct {
		const char *name;
		int value;
	} nv[] = {
		{ "apptype",	MAGIC_NO_CHECK_APPTYPE },
		{ "ascii",	MAGIC_NO_CHECK_ASCII },
		{ "compress",	MAGIC_NO_CHECK_COMPRESS },
		{ "elf",	MAGIC_NO_CHECK_ELF },
		{ "fortran",	MAGIC_NO_CHECK_FORTRAN },
		{ "soft",	MAGIC_NO_CHECK_SOFT },
		{ "tar",	MAGIC_NO_CHECK_TAR },
		{ "tokens",	MAGIC_NO_CHECK_TOKENS },
		{ "troff",	MAGIC_NO_CHECK_TROFF },
	};

#ifdef LC_CTYPE
	/* makes islower etc work for other langs */
	(void)setlocale(LC_CTYPE, "");
#endif

#ifdef __EMX__
	/* sh-like wildcard expansion! Shouldn't hurt at least ... */
	_wildcard(&argc, &argv);
#endif

	if ((progname = strrchr(argv[0], '/')) != NULL)
		progname++;
	else
		progname = argv[0];

	magicfile = default_magicfile;
	if ((usermagic = getenv("MAGIC")) != NULL)
		magicfile = usermagic;
	else
		if ((home = getenv("HOME")) != NULL) {
			if ((usermagic = malloc(strlen(home)
			    + sizeof(hmagic))) != NULL) {
				(void)strcpy(usermagic, home);
				(void)strcat(usermagic, hmagic);
				if (stat(usermagic, &sb)<0) 
					free(usermagic);
				else
					magicfile = usermagic;
			}
		}

#ifdef S_IFLNK
	flags |= getenv("POSIXLY_CORRECT") ? MAGIC_SYMLINK : 0;
#endif
#ifndef HAVE_GETOPT_LONG
	while ((c = getopt(argc, argv, OPTSTRING)) != -1)
#else
	while ((c = getopt_long(argc, argv, OPTSTRING, long_options,
	    &longindex)) != -1)
#endif
		switch (c) {
#ifdef HAVE_GETOPT_LONG
		case 0 :
			if (longindex == 1)
				help();
			break;
#endif
		case '0':
			nulsep = 1;
			break;
		case 'b':
			++bflag;
			break;
		case 'c':
			action = FILE_CHECK;
			break;
		case 'C':
			action = FILE_COMPILE;
			break;
		case 'd':
			flags |= MAGIC_DEBUG|MAGIC_CHECK;
			break;
		case 'e':
			for (i = 0; i < sizeof(nv) / sizeof(nv[0]); i++)
				if (strcmp(nv[i].name, optarg) == 0)
					break;

			if (i == sizeof(nv) / sizeof(nv[0]))
				errflg++;
			else
				flags |= nv[i].value;
			break;
			
		case 'f':
			if(action)
				usage();
			load(magicfile, flags);
			unwrap(optarg);
			++didsomefiles;
			break;
		case 'F':
			separator = optarg;
			break;
		case 'i':
			flags |= MAGIC_MIME;
			break;
		case 'k':
			flags |= MAGIC_CONTINUE;
			break;
		case 'm':
			magicfile = optarg;
			break;
		case 'n':
			++nobuffer;
			break;
		case 'N':
			++nopad;
			break;
#if defined(HAVE_UTIME) || defined(HAVE_UTIMES)
		case 'p':
			flags |= MAGIC_PRESERVE_ATIME;
			break;
#endif
		case 'r':
			flags |= MAGIC_RAW;
			break;
		case 's':
			flags |= MAGIC_DEVICES;
			break;
		case 'v':
			(void)fprintf(stdout, "%s-%d.%.2d\n", progname,
				       FILE_VERSION_MAJOR, patchlevel);
			(void)fprintf(stdout, "magic file from %s\n",
				       magicfile);
			return 1;
		case 'z':
			flags |= MAGIC_COMPRESS;
			break;
#ifdef S_IFLNK
		case 'L':
			flags |= MAGIC_SYMLINK;
			break;
		case 'h':
			flags &= ~MAGIC_SYMLINK;
			break;
#endif
		case '?':
		default:
			errflg++;
			break;
		}

	if (errflg) {
		usage();
	}

	switch(action) {
	case FILE_CHECK:
	case FILE_COMPILE:
		magic = magic_open(flags|MAGIC_CHECK);
		if (magic == NULL) {
			(void)fprintf(stderr, "%s: %s\n", progname,
			    strerror(errno));
			return 1;
		}
		c = action == FILE_CHECK ? magic_check(magic, magicfile) :
		    magic_compile(magic, magicfile);
		if (c == -1) {
			(void)fprintf(stderr, "%s: %s\n", progname,
			    magic_error(magic));
			return -1;
		}
		return 0;
	default:
		load(magicfile, flags);
		break;
	}

	if (optind == argc) {
		if (!didsomefiles) {
			usage();
		}
	}
	else {
		int i, wid, nw;
		for (wid = 0, i = optind; i < argc; i++) {
			nw = file_mbswidth(argv[i]);
			if (nw > wid)
				wid = nw;
		}
		for (; optind < argc; optind++)
			process(argv[optind], wid);
	}

	magic_close(magic);
	return magic->haderr ? -1 : 0;
}


private void
/*ARGSUSED*/
load(const char *m __unused, int flags)
{
	if (magic)
		return;
	magic = magic_open(flags);
	if (magic == NULL) {
		(void)fprintf(stderr, "%s: %s\n", progname, strerror(errno));
		exit(1);
	}
	if (magic_load(magic, magicfile) == -1) {
		(void)fprintf(stderr, "%s: %s\n",
		    progname, magic_error(magic));
		exit(1);
	}
}

/*
 * unwrap -- read a file of filenames, do each one.
 */
private void
unwrap(char *fn)
{
	char buf[MAXPATHLEN];
	FILE *f;
	int wid = 0, cwid;
	size_t len;

	if (strcmp("-", fn) == 0) {
		f = stdin;
		wid = 1;
	} else {
		if ((f = fopen(fn, "r")) == NULL) {
			(void)fprintf(stderr, "%s: Cannot open `%s' (%s).\n",
			    progname, fn, strerror(errno));
			exit(1);
		}

		while (fgets(buf, MAXPATHLEN, f) != NULL) {
			len = strlen(buf);
			if (len > 0 && buf[len - 1] == '\n')
				buf[len - 1] = '\0';
			cwid = file_mbswidth(buf);
			if (cwid > wid)
				wid = cwid;
		}

		rewind(f);
	}

	while (fgets(buf, MAXPATHLEN, f) != NULL) {
		len = strlen(buf);
		if (len > 0 && buf[len - 1] == '\n')
			buf[len - 1] = '\0';
		process(buf, wid);
		if(nobuffer)
			(void)fflush(stdout);
	}

	(void)fclose(f);
}

/*
 * Called for each input file on the command line (or in a list of files)
 */
private void
process(const char *inname, int wid)
{
	const char *type;
	int std_in = strcmp(inname, "-") == 0;

	if (wid > 0 && !bflag) {
		(void)printf("%s", std_in ? "/dev/stdin" : inname);
		if (nulsep)
			(void)putc('\0', stdout);
		else
			(void)printf("%s", separator);
		(void)printf("%*s ",
		    (int) (nopad ? 0 : (wid - file_mbswidth(inname))), "");
	}

	type = magic_file(magic, std_in ? NULL : inname);
	if (type == NULL)
		(void)printf("ERROR: %s\n", magic_error(magic));
	else
		(void)printf("%s\n", type);
}


#if 0
/*
 * byteconv4
 * Input:
 *	from		4 byte quantity to convert
 *	same		whether to perform byte swapping
 *	big_endian	whether we are a big endian host
 */
private int
byteconv4(int from, int same, int big_endian)
{
	if (same)
		return from;
	else if (big_endian) {		/* lsb -> msb conversion on msb */
		union {
			int i;
			char c[4];
		} retval, tmpval;

		tmpval.i = from;
		retval.c[0] = tmpval.c[3];
		retval.c[1] = tmpval.c[2];
		retval.c[2] = tmpval.c[1];
		retval.c[3] = tmpval.c[0];

		return retval.i;
	}
	else
		return ntohl(from);	/* msb -> lsb conversion on lsb */
}

/*
 * byteconv2
 * Same as byteconv4, but for shorts
 */
private short
byteconv2(int from, int same, int big_endian)
{
	if (same)
		return from;
	else if (big_endian) {		/* lsb -> msb conversion on msb */
		union {
			short s;
			char c[2];
		} retval, tmpval;

		tmpval.s = (short) from;
		retval.c[0] = tmpval.c[1];
		retval.c[1] = tmpval.c[0];

		return retval.s;
	}
	else
		return ntohs(from);	/* msb -> lsb conversion on lsb */
}
#endif

size_t
file_mbswidth(const char *s)
{
#if defined(HAVE_WCHAR_H) && defined(HAVE_MBRTOWC) && defined(HAVE_WCWIDTH)
	size_t bytesconsumed, old_n, n, width = 0;
	mbstate_t state;
	wchar_t nextchar;
	(void)memset(&state, 0, sizeof(mbstate_t));
	old_n = n = strlen(s);

	while (n > 0) {
		bytesconsumed = mbrtowc(&nextchar, s, n, &state);
		if (bytesconsumed == (size_t)(-1) ||
		    bytesconsumed == (size_t)(-2)) {
			/* Something went wrong, return something reasonable */
			return old_n;
		}
		if (s[0] == '\n') {
			/*
			 * do what strlen() would do, so that caller
			 * is always right
			 */
			width++;
		} else
			width += wcwidth(nextchar);

		s += bytesconsumed, n -= bytesconsumed;
	}
	return width;
#else
	return strlen(s);
#endif
}

private void
usage(void)
{
	(void)fprintf(stderr, USAGE, progname, progname);
#ifdef HAVE_GETOPT_LONG
	(void)fputs("Try `file --help' for more information.\n", stderr);
#endif
	exit(1);
}

#ifdef HAVE_GETOPT_LONG
private void
help(void)
{
	(void)puts(
"Usage: file [OPTION]... [FILE]...\n"
"Determine file type of FILEs.\n"
"\n"
"  -m, --magic-file LIST      use LIST as a colon-separated list of magic\n"
"                               number files\n"
"  -z, --uncompress           try to look inside compressed files\n"
"  -b, --brief                do not prepend filenames to output lines\n"
"  -c, --checking-printout    print the parsed form of the magic file, use in\n"
"                               conjunction with -m to debug a new magic file\n"
"                               before installing it\n"
"  -e, --exclude              exclude test from the list of test to be\n"
"                               performed for file. Valid tests are:\n"
"                               ascii, apptype, elf, compress, soft, tar\n"
"  -f, --files-from FILE      read the filenames to be examined from FILE\n"
"  -F, --separator string     use string as separator instead of `:'\n"
"  -i, --mime                 output mime type strings\n"
"  -k, --keep-going           don't stop at the first match\n"
"  -L, --dereference          causes symlinks to be followed\n"
"  -n, --no-buffer            do not buffer output\n"
"  -N, --no-pad               do not pad output\n"
"  -p, --preserve-date        preserve access times on files\n"
"  -r, --raw                  don't translate unprintable chars to \\ooo\n"
"  -s, --special-files        treat special (block/char devices) files as\n"
"                             ordinary ones\n"
"      --help                 display this help and exit\n"
"      --version              output version information and exit\n"
);
	exit(0);
}
#endif
